### DISCUSSION TITLE
Migrating to 2.0 with plain javascript window.__TAURI__.tauri undefined

### DISCUSSION THREAD

**Discussion by 0xflux** [tauri-apps/tauri#11511] [Q&A] [answered]
In the v1 documentation it lists a way of injecting the Tauri (runtime?) into the web view via:

```javascript
const { invoke } = window.__TAURI__.tauri
```

However; after trying to migrate to v2, in the web inspector I get the following error:

Uncaught TypeError: Cannot destructure property 'invoke' of 'window.__TAURI__.tauri' as it is undefined.

And as such, none of my callbacks or events will work.

Can plain old JavaScript be used for developing apps in v2? I'm coming to this not from a strong web background, and as such don't know frameworks such as Vue, Next, Vite, Svelt etc.

I can see in the v2 docs it lists withGlobalTauri as an attribute, so I assume the plain JavaScript development environment are supported?

My config (I'm just trying to use html and js to render a frontend, no fancy js servers or anything like that):

```json
{
  "build": {
    "frontendDist": "../ui"
  },
  "app": {
    "withGlobalTauri": true,
    "security": {
      "csp": null
    },
    "windows": [
      {
        "fullscreen": false,
        "height": 600,
        "resizable": true,
        "title": "Sanctum",
        "width": 1200
      }
    ]
  },
  "bundle": {
    "active": true,
    "category": "DeveloperTool",
    "copyright": "",
    "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ],
      "windows": {
        "certificateThumbprint": null,
        "digestAlgorithm": "sha256",
        "timestampUrl": ""
      }
  },
  "productName": "Sanctum",
  "version": "0.1.0",
  "identifier": "com.sanctum.app"
}
```

My branch for upgrading to 2.0 is https://github.com/0xflux/Sanctum/tree/upgrade-2-0/gui/src-tauri

Link to affected html file (ignore my horrible javascript skills) https://github.com/0xflux/Sanctum/blob/upgrade-2-0/gui/ui/antivirus.html

Many thanks

**Comment by Donald2010** [comment]
Don't know the reason either, but you can try using import instead.

**Comment by 0xflux** [comment, marked as answer]
Fixed with updated syntax below:

```javascript
const { invoke } = window.__TAURI__.core;
const { listen } = window.__TAURI__.event;
```

The only other addition i have is the capabilities > default.json:

```json
{
    "$schema": "./schemas/desktop-schema.json",
    "identifier": "main-capability",
    "description": "Capability for the main window",
    "windows": ["main"],
    "permissions": [
      "core:path:default",
      "core:event:default",
      "core:window:default",
      "core:app:default",
      "core:resources:default",
      "core:menu:default",
      "core:tray:default",
      "core:window:allow-set-title"
    ]
  }
```
